fendermint/testing/materializer/src/materializer.rs

```diff
@@ -0,0 +1,246 @@
+// Copyright 2022-2024 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+use async_trait::async_trait;
+use ethers::types::H160;
+use fvm_shared::econ::TokenAmount;
+use std::collections::BTreeMap;
+use tendermint_rpc::Url;
+
+use fendermint_vm_genesis::Collateral;
+
+use crate::{
+    manifest::Balance, AccountName, NodeName, RelayerName, ResourceHash, SubnetName, TestnetName,
+};
+
+/// The materializer is a component to provision resources of a testnet, and
+/// to carry out subsequent commands on them, e.g. to restart nodes.
+///
+/// By contrast, the role of the [Testnet] is to keep related items organised
+/// and accessible for the integration tests, carrying out the operations with
+/// the help of the materializer, which should keep the [Testnet] itself testable.
+///
+/// The materializer might not actually instantiate the resources. By returning
+/// abstract types instead of concrete values, it is possible to just collect the
+/// operations and use them to validate the behaviour of whatever is driving
+/// the materializer. We can use this for dry-runs as well.
+///
+/// A live materializer should persist its logs, so that it can be resumed.
+/// For example we can create an run a testnet externally, then parse the manifest
+/// and the materializer logs inside a test to talk to one of the nodew, and the
+/// materializer should be able to return to the test correct JSON-RPC endpoints.
+///
+/// Some of the operations of the materializer should be idempotent, e.g. the
+/// creation of a wallet or a node should only happen once.
+///
+/// The types returned might have their own logic to execute when dropped, to free
+/// resources. This might happen only if the resource is not an externally managed
+/// one, e.g. a testnet set up before tests are run, which the materializer should
+/// know.
```

Comment:
So the resource returned might (if applicable) implement the Drop trait for RAII so we properly free the underlying resource (docker container, process, etc)

Reply:
Yes, and example of that is in https://github.com/consensus-shipyard/ipc/pull/700/files#diff-e53ae3b1e30b37699016a7e43c4462b00ece2c8803212e253d6f3743f09abb96